// overlayscrollbars-react/overlayscrollbars-react.es.js
import b, { useMemo as w, useRef as f, useEffect as v, forwardRef as C, useImperativeHandle as E } from "react";
import { OverlayScrollbars as R } from "overlayscrollbars";
const O = () => {
  if (typeof window > "u") {
    const n = () => {
    };
    return [n, n];
  }
  let l, c;
  const t = window, o = typeof t.requestIdleCallback == "function", a = t.requestAnimationFrame, i = t.cancelAnimationFrame, r = o ? t.requestIdleCallback : a, u = o ? t.cancelIdleCallback : i, s = () => {
    u(l), i(c);
  };
  return [
    (n, e) => {
      s(), l = r(
        o ? () => {
          s(), c = a(n);
        } : n,
        typeof e == "object" ? e : { timeout: 2233 }
      );
    },
    s
  ];
}, S = (l) => {
  const { options: c, events: t, defer: o } = l || {}, [a, i] = w(O, []), r = f(null), u = f(o), s = f(c), n = f(t);
  return v(() => {
    u.current = o;
  }, [o]), v(() => {
    const { current: e } = r;
    s.current = c, R.valid(e) && e.options(c || {}, !0);
  }, [c]), v(() => {
    const { current: e } = r;
    n.current = t, R.valid(e) && e.on(t || {}, !0);
  }, [t]), v(
    () => () => {
      var e;
      i(), (e = r.current) == null || e.destroy();
    },
    []
  ), w(
    () => [
      (e) => {
        const y = r.current;
        if (R.valid(y))
          return;
        const d = u.current, I = s.current || {}, m = n.current || {}, p = () => r.current = R(e, I, m);
        d ? a(p, d) : p();
      },
      () => r.current
    ],
    []
  );
}, F = (l, c) => {
  const { element: t = "div", options: o, events: a, defer: i, children: r, ...u } = l, s = t, n = f(null), e = f(null), [y, d] = S({ options: o, events: a, defer: i });
  return v(() => {
    const { current: I } = n, { current: m } = e;
    return I && m && y({
      target: I,
      elements: {
        viewport: m,
        content: m
      }
    }), () => {
      var p;
      return (p = d()) == null ? void 0 : p.destroy();
    };
  }, [y, t]), E(
    c,
    () => ({
      osInstance: d,
      getElement: () => n.current
    }),
    []
  ), /* @__PURE__ */ b.createElement(s, {
    "data-overlayscrollbars-initialize": "",
    ref: n,
    ...u
  }, /* @__PURE__ */ b.createElement("div", {
    ref: e
  }, r));
}, A = C(F);
export {
  A as OverlayScrollbarsComponent,
  S as useOverlayScrollbars
};
